internal/consensus/state.go

```diff
@@ -1615,20 +1615,14 @@ func (cs *State) enterPrecommit(height int64, round int32) {
 	}
 	// At this point, +2/3 prevoted for a particular block.
 
-	// If we never received a proposal for this block, we must precommit nil
-	if cs.Proposal == nil || cs.ProposalBlock == nil {
+	// If we did not receive a proposal block to vote for, we must precommit nil
+	// FIXME: although +2/3 prevoted for the block, therefore it is valid.
```

Comment:
Who's the FIXME label for? As stated in #2119, this is correct accordingly to the spec, what's there for the code monkeys to fix?

Reply:
Good point. I think the question here is that when adapting the code to match the spec we are losing some opportunities to, in a smart way, speed up decisions when we have Byzantine proposers, for example.

In our in production (current) releases, we can prevote and precommit to a block that matches our locked block. Because this is safe (at least, I don't see how it would not be safe). This code, for the precommit step is still here, btw, see lines 1633/1627.

When we require having `cs.ProposalBlock` to precommit a block, we are not using the possibility added to the above mentioned logic/lines.

In other words, what we would need here is a way to re-use the `LockedBlock`, setting `cs.Proposal` to it, in case the block we are looking for matches the locked one. This save us, in particular, the cost of downloading it again in a round `r > cs.LockedRound`. We may also consider the `cs.ValidRound`, for the same reasons. But this should be addressed by another issue/PR.

Reply:
If you navigate the history of commits, you realize that I removed this check altogether. The problem of which is that `TestStateLock_MissingProposalWhenPOLForLockedBlock` test fails. I then reintroduced this check to make this test to pass.

The fun part is that the description of the test does not match the test itself. Which is odd.

Reply:
Hmm, not sure, maybe the FIXME explanation should be expanded or placed elsewhere, to reflect on the missed opportunities and/or the oddly specified test, and the note should state what it does now.